woocommerce/woocommerce-ios
# [Mobile Payments] Show error when mandatory update fails because of low battery

Part of #5036 

This handles the case when a mandatory update fails to install because the reader has less than 50% battery.
Note that I've only been able to test this with the simulated readers, which only offers low battery failure simulation for mandatory updates. This doesn't address optional updates failing yet (they'd fail silently), or other update failures on mandatory updates (they would show up as a generic failure to connect).

When the charge % is known:

Light|Dark
-|-
![Simulator Screen Shot - iPhone 12 mini - 2021-09-28 at 12 48 33](https://user-images.githubusercontent.com/8739/135100132-3a5fef2e-5a5b-4c8a-8053-bc2cb847d1fc.png)|![Simulator Screen Shot - iPhone 12 mini - 2021-09-28 at 12 48 36](https://user-images.githubusercontent.com/8739/135100140-d0d96859-020b-4f6b-a911-9fe7fd05cdc8.png)

When the charge % is unknown (@woocommerce/mobile-designers the Figma file didn't include this scenario, so I tweaked the copy a little bit):

Light|Dark
-|-
![Simulator Screen Shot - iPhone 12 mini - 2021-09-28 at 12 48 05](https://user-images.githubusercontent.com/8739/135100241-92a08ad5-bd49-450b-9e52-96edc1778fe9.png)|![Simulator Screen Shot - iPhone 12 mini - 2021-09-28 at 12 48 17](https://user-images.githubusercontent.com/8739/135100246-693c81e0-9e77-495c-8ca5-10025f3ed54a.png)

## To test

1. On Xcode, edit the WooCommerce scheme, and enable `-simulate-stripe-card-reader` in the Arguments tab
2. In `StripeCardReaderService.start(_:)`, after the SDK has been initialized (I add this right after `Terminal.shared.delegate = self`), add the following line: `Terminal.shared.simulatorConfiguration.availableReaderUpdate = .lowBattery`
3. Run the app
4. Go to Settings > In-Person Payments > Manage Card Reader and connect to a reader
5. After the reader connects, it should automatically attempt to update and show the new error

Note that sometimes the simulated reader reports 85% battery (even with the simulated low battery error 🙃), and sometimes it's `nil`. Follow the steps 2-3 times and you should see the UI for both cases

Update release notes:
